Added Regex Constructors and Match Operator

ggetz opened the pull request:
Firstly, I added functions to replace all backlashes with an obscure string before using jsep, (I'm using `@#%`), and back after parsing. I added a test to make sure this is happening correctly.

I also added a constructor function to create a regex `RegExp()`. It requires either no arguments (which creates an empty regex `/(?:)/`) or a literal. If a literal is not received, an error is thrown. Like the javascript `RegExp()` constructor, it can take other literal types as arguments. For example, `RegExp(1)` is legal and evaluates to `/1/`. It will not take a color. I added tests to check for constructing with strings, numbers, and booleans.

Since JavaScript does not have an operator to indicate regex matching, I also used the `=~` operator to check regex matches, as @pierotofy suggested. It is treated as commutative, and return `true` if the regex matches and `false` otherwise. It throws an error if one of the arguments is not a regex. I added tests to check for these cases.

Once this PR is closed, I'll implement the `!~` operator, and an operator to that returns the first captured match.

pjcozzi commented:
> I also used the =~ operator to check regex matches

Did you consider using `test` and `exec` functions like the JavaScript `RegEx` object (example here: https://github.com/AnalyticalGraphicsInc/3d-tiles/issues/2#issuecomment-187728554)?  I think this fits better than adding an operator; that could come later if the syntactic sugar is needed.

ggetz commented:
I thought an operator would fit in better as we don't have any other functions that are performed using the  dot operator.

pjcozzi commented:
It can be a separate PR, but for constructing `Color` and `RegExp` objects, I'm afraid we have to require using `new` just like in JavaScript, e.g., `(new Color()) === (new Color())`.  As the styling language grows, this would fit.  We could also just drop the constructor functions, and have `color` and `regExp` functions that return new objects.  The expressions will be a bit cleaner, but it doesn't map as well to JavaScript (e.g., where it is `new RegExp`).

pjcozzi commented:
> I thought an operator would fit in better as we don't have any other functions that are performed using the dot operator.

That's not a good enough reason since we will, of course, have these in the future, e.g., `color.darken(0.5)`.

ggetz commented:
> It can be a separate PR, but for constructing Color and RegExp objects, I'm afraid we have to require using new just like in JavaScript

But we have the `Number()`, `Boolean()`, and `String()` functions, which convert to a literal. Aren't we doing something similar with the `Color()` function?

ggetz commented:
> That's not a good enough reason since we will, of course, have these in the future, e.g., color.darken(0.5).

OK, I will change it.

pjcozzi commented:
> But we have the Number(), Boolean(), and String() functions, which convert to a literal. Aren't we doing something similar with the Color() function since we are treating colors as a literal?

No, those are different.  Did you read the link I posted earlier today: http://stackoverflow.com/questions/4719320/new-number-vs-number

ggetz commented:
> No, those are different. Did you read the link I posted earlier today

Yes, I was asking since we are treating colors as a literal type, like Number or Boolean or String.

ggetz commented:
Also in javascript (new Object()) === (new Object()) is not true. If we use `new` for `Color`, should  (new Color()) === (new Color()) behave differently?

ggetz commented:
In addition, if we do end up using bot `new` and `.test()` for regex's, then in order to use them, we'll always have to use `(new RegExp()).test()`. Are we OK with that?

ggetz commented:
OK, I removed the operator and added the "function" `test`.

pjcozzi commented:
`Number(1)` is a cast.

`new Number(1)` constructs a container object.

Look at them in the console.

In our case, `(new Color()) === (new Color())` would be true because we overloaded the `===` operator so `Color` feels like a native type.

> In addition, if we do end up using bot new and .test() for regex's, then in order to use them, we'll always have to use (new RegExp()).test(). Are we OK with that?

This is, of course, the pain point.  If/when the styling language grows, these types of things are going to happen since it will adopt more and more JavaScript.

Given how ugly `new` is for our initial use cases, we could also have helper creation functions, e.g., `color` and `regExp`, that do not require new.  This is similar to `createXXX` functions in Cesium.  Their implementation would be, for example, `return new Color(arguments);`.

We should not keep it as is since we have a constructor function convention, e.g., `Color`, that does not require `new`, which JavaScript doesn't do anywhere AFAIK.

Let's discuss tomorrow.

pjcozzi commented on Source/Scene/Expression.js:
```diff
@@ -23,9 +29,17 @@ define([
+        }
+        //>>includeEnd('debug');
+
+        jsep.addBinaryOp("=~", 0);
```
Should this code still be here?

ggetz replied:
No, I will remove it.

pjcozzi commented on Source/Scene/Expression.js:
```diff
@@ -112,10 +134,33 @@ define([
+            call = ast.callee.property.name;
+            if (call === 'test') {
+                if (ast.callee.object.callee.name !== 'RegExp') {
+                    //>>includeStart('debug', pragmas.debug);
```
Put this comment around the `if` statement.

pjcozzi commented on Source/Scene/Expression.js:
```diff
@@ -263,10 +319,7 @@ define([
-                op === '/' || op === '%' || op === '===' ||
-                op === '!==' || op === '>' || op === '>=' ||
-                op === '<' || op === '<=') {
+            if (binaryOperators.indexOf(op) > -1) {
```
This may have hurt "compiling" performance, which will be important for the responsive of a UI generating styles.  I don't know for sure so we can see what a profile looks like later; at the least, we can order the linear search so the most common operators are first.

ggetz replied:
So should I leave it as is for now?

pjcozzi replied:
Yes, I just mentioned it because it jumped out at me for some reason.

pjcozzi commented:
What do you think about supporting the flags argument to the RegEx constructor: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp

ggetz commented:
> What do you think about supporting the flags argument to the RegEx constructor

I think that's fine. I'll put it in.

ggetz commented:
I made the fixes and added the optional flags argument, as well as added tests to check.

pjcozzi commented on Source/Scene/Expression.js:
```diff
@@ -198,6 +242,26 @@ define([
+            }
+            val = args[0];
+            if (args.length > 1) {
+                try {
```
What could throw here?  Just the `RegExp` constructor?  If so, separate that out into its own block.

pjcozzi commented:
I think we are going to need `exec` right away (perhaps just return `[0]` for now?), for example, see https://github.com/AnalyticalGraphicsInc/3d-tiles/issues/2#issuecomment-187728554

pjcozzi commented on Specs/Scene/ExpressionSpec.js:
```diff
@@ -681,6 +686,68 @@ defineSuite([
+        }).toThrowDeveloperError();
+    });
+
+    it('evaluates regex matches operator', function() {
```
We don't need to overdo it since we can trust the JavaScript RegExp engine, but I think this needs a couple more non-trivial tests with more complicated regular expressions and compound expressions.

pjcozzi commented:
That's all my comments.  Also update the spec.  We can do the constructor function changes we discussed offline in a separate PR.

ggetz commented:
> I think we are going to need exec right away (perhaps just return [0] for now?)

Yes, that's how I had planned on implementing it. I should include it in this PR?

pjcozzi commented:
Yes.

ggetz commented:
OK, I added the exec function (which returns `[1]` because` [0]` is the entire matched string, not the first captured result).

I also added a less trivial test for both `test` and `exec` which uses an example from the MDN documentation.

pjcozzi commented:
:+1:
